### ISSUE TITLE
Jax at head infinite loops printing an array

### ISSUE THREAD

**Issue by stellaraccident** [google/jax#14422]
### Description

Tensorflow checked out commit: d06d16ef29767129aeaaefd007481aeb759e5400
Jax: dc6bf9b725fcdd96f8152d6d3a1ce3f0c14d9ced

This simple test program:

```
import jax

# Do once and print.
a = jax.numpy.asarray([1, 2, 3, 4, 5, 6, 7, 8, 9])
b = a
for i in range(100):
  b = jax.numpy.asarray([i]) * a + b
print(b)
```

Fails with:

```
...
  File "/usr/local/google/home/laurenzo/src/openxla-pjrt-plugin/external/.clones/jax/jax/_src/array.py", line 351, in __array__
    return np.asarray(self._value, dtype=dtype)
  File "/usr/local/google/home/laurenzo/src/openxla-pjrt-plugin/external/.clones/jax/jax/_src/array.py", line 497, in _value
    self._npy_value = np.asarray(self._arrays[0])  # type: ignore
  File "/usr/local/google/home/laurenzo/src/openxla-pjrt-plugin/external/.clones/jax/jax/_src/array.py", line 351, in __array__
    return np.asarray(self._value, dtype=dtype)
  File "/usr/local/google/home/laurenzo/src/openxla-pjrt-plugin/external/.clones/jax/jax/_src/array.py", line 496, in _value
    if self.is_fully_replicated:
  File "/usr/local/google/home/laurenzo/src/openxla-pjrt-plugin/external/.clones/jax/jax/_src/array.py", line 325, in is_fully_replicated
    return self.shape == self._arrays[0].shape
  File "/usr/local/google/home/laurenzo/src/openxla-pjrt-plugin/external/.clones/jax/jax/_src/array.py", line 196, in shape
    return self.aval.shape
RecursionError: maximum recursion depth exceeded while calling a Python object
```

I can reproduce this on all backends, including CPU.

**Comment by mattjj** [comment]
Hey @stellaraccident !

It works on my machine. That is, both at https://github.com/google/jax/commit/dc6bf9b725fcdd96f8152d6d3a1ce3f0c14d9ced and at current github HEAD, using pypi's latest jaxlib (but not a local build), that script doesn't fail.

Did you build jaxlib yourself, or install it? Can you try it with pypi's latest? (I think: `pip install --upgrade jaxlib`.)

I wonder if it's a jax/jaxlib version incompatibility issue...

**Comment by stellaraccident** [comment]
> I wonder if it's a jax/jaxlib version incompatibility issue...

I suspect it is. I'm building with a jaxlib built from head. Working on some features that need the latest, and thought it was something I did, so I reproduced it with a clean build of at head jax and tensorflow repos and a custom build jaxlib.

There isn't a problem with what is on pypi and I haven't fully bisected but this just started recently.

**Comment by mattjj** [comment]
Looks like cl/508789203 is a rollback in progress!

**Comment by yashk2810** [comment]
I think this PR rollback should fix it: https://github.com/google/jax/pull/14421

**Comment by stellaraccident** [comment]
Not so sure: Just rebased to head (61da7811742d69b3d0e33c211418d1d131cf475b) and it appears to be working. But if I cherry-pick that rollback, the issue returns. So something fixed it in the last few hours but it looks like the rollback might re-break it :)

In any case, going to let it be for the weekend and hope it stabilizes...

**Comment by yashk2810** [comment]
I think it is a combination of jax at head and TF at head (jaxlib). We need both changes to be reverted to get the previous behavior.

We want to rollback that change anyways because of other failures in google. Let's wait for the rollback to get in and then try again? (Or i can by building jaxlib from scratch)

**Comment by stellaraccident** [comment]
(I'm not blocked now and assume that you all can get it past the oscillation... the repro is pretty straight-forward)

**Comment by yashk2810** [comment]
Rollback is submitted and is OSS'd

JAX commit: https://github.com/google/jax/commit/9316188b3af802c136e78bc1908a46992633e1f0

TF commit: https://github.com/tensorflow/tensorflow/commit/1be8ca3a774653e69831fe82cf4b140062dcfcf6

I am building and reproing now

**Comment by yashk2810** [comment]
The rollback fixes it

Command I used to build jaxlib: `python3 build/build.py --bazel_options=--override_repository=org_tensorflow=/home/yashkatariya/tensorflow`. The tensorflow directory was cloned from master branch of TF. Same for JAX (main branch of JAX)

```
>>> import jax
>>> jax.devices()
No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
[CpuDevice(id=0)]
>>> jax.__version__
'0.4.4'
>>> import jaxlib
>>> jaxlib.__version__
'0.4.4'
>>> a = jax.numpy.asarray([1, 2, 3, 4, 5, 6, 7, 8, 9])
>>> b = a
>>> for i in range(100):
...     b = jax.numpy.asarray([i]) * a + b
...
>>> print(b)
[ 4951  9902 14853 19804 24755 29706 34657 39608 44559]
>>>
```

**Comment by yashk2810** [comment]
@stellaraccident feel free to try it out at HEAD of both jax and TF and see if it is resolved in your env too :)

**Comment by stellaraccident** [comment]
Thanks! I'll check tomorrow. FWIW, here is the environment I am verifying: https://github.com/openxla/openxla-pjrt-plugin

That should work tomorrow (after a nightly dep roll) with just a local patch to work around: https://github.com/openxla/xla/issues/1237 (shameless plug to get some attention on that too :) )

**Comment by pschuh** [comment]
Hi Stella, can you expound more on what is going on here? Does this mean that you're using the newest python but an old build of the jaxlib extention? Are you using a pjrt version of iree?

**Comment by stellaraccident** [comment]
Sure. I checked out the jax repo from head then synced the backing tensorflow to near-head and built a jaxlib per https://jax.readthedocs.io/en/latest/developer.html#building-jaxlib-from-source-with-a-modified-tensorflow-repository. This was because I am working with some folks who have landed changes to jaxlib/tensorflow that I need in order to make progress. Due to this, I'm often developing with a near-head jaxlib.

I am doing this to support the PJRT plugin, which is where I noticed the issue. However, this bug report is just the result of running the script I included against the default XLA CPU backend, so it is not related to anything I am doing.

**Comment by pschuh** [comment]
Right, but we have a tensorflow/compiler/xla/python/xla_client.py with a _version number which allows us to synchronize changes between jax and jaxlib. If that file gets out of sync with the compiled blob, then you could get problems like this.

**Comment by stellaraccident** [comment]
Is there a written procedure for upgrading jaxlib? When developing near-head on the OSS side, it seems like step 1 might be to do that locally and then work further. All I can see is this: https://github.com/google/jax/blob/c49af18b9b6a560c1d6cbc516e5c800ad4fb2ca1/WORKSPACE#L3

Which implies that we always have our version numbers and such in a state so that just bumping the tf hash should be sufficient (modulo bugs and other stuff, I assume). Let's say that instead of encountering this issue while trying to accomplish my own objective, my objective was to upgrade the tensorflow hash. I would have updated the pin as at the top of the file and then encountered this error, which presumably would have needed to be fixed prior to actually bumping the pin.
